src/function/cast/struct_cast.cpp

```diff
@@ -229,6 +229,120 @@ static bool StructToVarcharCast(Vector &source, Vector &result, idx_t count, Cas
+	vector<BoundCastInfo> value_casts;
+	auto target_value_type = MapType::ValueType(target);
+	for (idx_t i = 0; i < StructType::GetChildCount(source); i++) {
+		auto value_cast = input.GetCastFunction(StructType::GetChildType(source, i), target_value_type);
+		value_casts.push_back(std::move(value_cast));
+	}
+	return make_uniq<StructToMapBoundCastData>(std::move(key_cast), std::move(value_casts));
+}
+
+unique_ptr<FunctionLocalState>
+StructToMapBoundCastData::InitStructToMapCastLocalState(CastLocalStateParameters &parameters) {
+	auto &cast_data = parameters.cast_data->Cast<StructToMapBoundCastData>();
+	auto result = make_uniq<StructToMapCastLocalState>();
+
+	if (cast_data.key_cast.init_local_state) {
+		CastLocalStateParameters child_params(parameters, cast_data.key_cast.cast_data);
+		result->key_state = cast_data.key_cast.init_local_state(child_params);
+	}
+
+	for (auto &entry : cast_data.value_casts) {
+		unique_ptr<FunctionLocalState> child_state;
+		if (entry.init_local_state) {
+			CastLocalStateParameters child_params(parameters, entry.cast_data);
+			child_state = entry.init_local_state(child_params);
+		}
+		result->value_states.push_back(std::move(child_state));
+	}
+	return std::move(result);
+}
+
+static bool StructToMapCast(Vector &source, Vector &result, idx_t count, CastParameters &parameters) {
+
+	if (source.GetVectorType() == VectorType::CONSTANT_VECTOR) {
+		// If the source is a constant vector the result may be as well
+		result.SetVectorType(VectorType::CONSTANT_VECTOR);
+		// No need to do anything with null structs
+		if (ConstantVector::IsNull(source)) {
+			ConstantVector::SetNull(result, true);
+			return true;
+		}
+	}
+
+	auto &cast_data = parameters.cast_data->Cast<StructToMapBoundCastData>();
+	auto &local_state = parameters.local_state->Cast<StructToMapCastLocalState>();
+
+	// Grab all struct columns
+	auto &source_children = StructVector::GetEntries(source);
+	idx_t field_count = source_children.size();
+	idx_t total_count = count * field_count;
+
+	// Allocate result
+	ListVector::Reserve(result, total_count);
+
+	// Create key vector with VARCHAR keys (could make this a dictionary vector...?)
+	Vector varchar_keys(LogicalType::VARCHAR, total_count);
+	auto key_data = FlatVector::GetData<string_t>(varchar_keys);
+	auto &field_types = StructType::GetChildTypes(source.GetType());
+	for (idx_t row_idx = 0; row_idx < count; row_idx++) {
+		for (idx_t field_idx = 0; field_idx < field_count; field_idx++) {
+			auto global_idx = row_idx * field_count + field_idx;
+			auto &field_name = field_types[field_idx].first;
+			key_data[global_idx] = StringVector::AddString(varchar_keys, field_name);
+		}
+	}
+
+	// Cast keys to result
+	auto &map_keys = MapVector::GetKeys(result);
+	CastParameters key_parameters(parameters, cast_data.key_cast.cast_data, local_state.key_state);
+	auto keys_converted = cast_data.key_cast.function(varchar_keys, map_keys, total_count, key_parameters);
```

Comment:
In the local state creation we checked if it had an `init_local_state`, but here we assume that a local state was created

Either having a `init_local_state` should be an assertion, so we can assume it's present here
Or we need to be prepared for it being missing here

Reply:
You mean when we pass `local_state.key_state` as param when `key_parameters` is created? `local_state` is a `optional_ptr<FunctionLocalState>` in `CastParameters`, my thinking was that it's allowed to be null, so I wouldn't need to check? Or (not unlikely 😅 ) am I missing something?